File: solutions/java/dnd-character/1/src/main/java/DnDCharacter.java
import java.util.Arrays;
import java.util.List;
import java.util.Random;
import java.util.random.*;
import java.util.stream.IntStream;

public class DnDCharacter {
    private int strength;
    private int dexerity;
    private int constitution;
    private int intelligence;
    private int wisdom;
    private int charisma;

    public DnDCharacter(){
        strength = ability(rollDice());
        dexerity = ability(rollDice());
        constitution = ability(rollDice());
        intelligence = ability(rollDice());
        wisdom = ability(rollDice());
        charisma = ability(rollDice());
    }

    public int ability(List<Integer> scores) {
        return scores.stream()
            .mapToInt(Integer::intValue)
            .sorted()
            .skip(1)
            .sum();
    }

    public List<Integer> rollDice() {
        Random r = new Random();
        return Arrays.stream(IntStream.generate(() -> r.nextInt(1, 7))
                             .limit(4).toArray())
            .boxed()
            .toList();
    }

    public int modifier(int input) {
        return (int)Math.floor((input - 10) / 2.0);
    }

    public int getStrength() {
        return strength;
    }

    public int getDexterity() {
        return dexerity;
    }

    public int getConstitution() {
        return constitution;
    }

    public int getIntelligence() {
        return intelligence;
    }

    public int getWisdom() {
        return wisdom;
    }

    public int getCharisma() {
        return charisma;
    }

    public int getHitpoints() {
        return 10 + modifier(getConstitution());
    }
}

File: solutions/c/dnd-character/3/dnd_character.c
#include "dnd_character.h"

#include <limits.h>
#include <stdlib.h>
#include <math.h>

int ability(void) {
    return rand() % 15 + 3;
}

int modifier(int score){
    return floor((score - 10) / 2.);
}

dnd_character_t make_dnd_character(void) {
    dnd_character_t character;
    character.strength     = ability(),
    character.strength     = ability(),
    character.dexterity    = ability(),
    character.constitution = ability(),
    character.intelligence = ability(),
    character.wisdom       = ability(),
    character.charisma     = ability(),
    character.hitpoints    = 10 + modifier(character.constitution);
    return character;
}